databricks/koalas/base.py

```diff
@@ -57,6 +57,12 @@ def wrapper(self, *args):
             args = [arg._scol if isinstance(arg, IndexOpsMixin) else arg for arg in args]
             scol = f(self._scol, *args)
 
+            # If f is a logistic operator, fill NULL with False
+            log_ops = ['eq', 'ne', 'lt', 'le', 'ge', 'gt']
+            is_log_op = any(f == getattr(spark.Column, f'__{log_op}__') for log_op in log_ops)
+            if is_log_op:
+                scol = F.when(scol.isNull(), False).otherwise(scol)
+
```

Comment:
@HyukjinKwon @ueshin (cc @itholic @charlesdong1991 )
Not sure if this is the right implementation ...

Reply:
maybe `scol = F.when(scol.isNull(), False).otherwise(True)` ?

Reply:
Ah, and i think maybe it should be better to make test of this if you available  😄

Reply:
@itholic Thanks! I haven't fixed the tests yet because I want to make sure first if this is the right implementation.

Reply:
Actually the problem is not that simple. E.g.,:

```py
>>> pser == 0
0     True
1    False
2    False
3    False
dtype: bool
>>> pser != 0
0    False
1     True
2     True
3     True
dtype: bool
```

and

```py
>>> ks.from_pandas(pser) == 0
0     True
1    False
2    False
3     None
Name: 0, dtype: object
>>> ks.from_pandas(pser) != 0
0    False
1     True
2     True
3     None
Name: 0, dtype: object
```

If we simply convert `None` to `False`, the `!=` case is not the same as pandas'.

We might need to add some logic one-by-one.

Reply:
my pleasure :)

Reply:
@ueshin 
ah, it looks way complicated than i thought. 😅 
thanks for the comment!

Reply:
@ueshin Thanks! Is there any case where the current way doesn't work?

Reply:
Speaking of the behavior difference,

```py
>>> ~pd.Series([True, False, np.nan])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/ueshin/workspace/databricks-koalas/miniconda/envs/databricks-koalas_3.6/lib/python3.6/site-packages/pandas/core/generic.py", line 1465, in __invert__
    arr = operator.inv(com.values_from_object(self))
TypeError: bad operand type for unary ~: 'float'
>>> ~pd.Series([True, False, None])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/ueshin/workspace/databricks-koalas/miniconda/envs/databricks-koalas_3.6/lib/python3.6/site-packages/pandas/core/generic.py", line 1465, in __invert__
    arr = operator.inv(com.values_from_object(self))
TypeError: bad operand type for unary ~: 'NoneType'
```

whereas:

```py
>>> ~ks.from_pandas(pd.Series([True, False, np.nan]))
0    False
1     True
2     None
Name: 0, dtype: object
>>> ~ks.from_pandas(pd.Series([True, False, None]))
0    False
1     True
2     None
Name: 0, dtype: object
```

I'm not sure we should raise an error for this.

Reply:
This is funny.

```py
>>> pd.Series([True, True, np.nan]) | pd.Series([True, np.nan, True])
0     True
1     True
2    False
dtype: bool
```

Reply:
hm, let's focus only `['eq', 'ne', 'lt', 'le', 'ge', 'gt']` here.
I roughly checked the behavior and seems like `ne` is only an exception.

Reply:
> This is funny.
> 
> ```python
> >>> pd.Series([True, True, np.nan]) | pd.Series([True, np.nan, True])
> 0     True
> 1     True
> 2    False
> dtype: bool
> ```
> 
> Why 1 and 2 are different result..

This should be reported to pandas 😅

Reply:
hmm

```py
>>> pser < np.nan
0    False
1    False
2    False
3    False
dtype: bool
>>> pser > np.nan
0    False
1    False
2    False
3    False
dtype: bool
>>>
>>> pser < None
0    False
1    False
2    False
3    False
dtype: bool
>>> pser > None
0    False
1    False
2    False
3    False
dtype: bool
```

whereas

```py
>>> ks.from_pandas(pser) < np.nan
0    True
1    True
2    True
3    None
Name: 0, dtype: object
>>> ks.from_pandas(pser) > np.nan
0    False
1    False
2    False
3     None
Name: 0, dtype: object
>>> ks.from_pandas(pser) < None
0    None
1    None
2    None
3    None
Name: 0, dtype: object
>>> ks.from_pandas(pser) > None
0    None
1    None
2    None
3    None
Name: 0, dtype: object
```

Reply:
> This is funny.
> 
> ```python
> >>> pd.Series([True, True, np.nan]) | pd.Series([True, np.nan, True])
> 0     True
> 1     True
> 2    False
> dtype: bool
> ```
> 
> Why 1 and 2 are different result..

i think this is because python, `nan's` don't compare equal, and pandas uses this fact, so it will turn `False` no matter what you compare with `np.nan`, so in this case, 2 will return `False`, the 1 will return `True` is because, in `|` operator in Python, if the stuff on the left is already `True`, python will directly return `True` and not continue to compare with the stuff on the right of `|`.

Reply:
@charlesdong1991 
Does Python `|` operator work that way?

Reply:
@charlesdong1991 I wrote a simple test.
```python
import time

def sleep_10_seconds():
  time.sleep(10)
  return False

# does evaluates sleep_10_seconds
if True | sleep_10_seconds():
  print('yes')

# doesn't evaluates sleep_10_seconds
if True or sleep_10_seconds():
  print('yes')
```

Reply:
@harupy i believe so.

the `|` operator will return the first object that evaluates to true or the last object in the expression, regardless of its truth value.

Reply:
oh, i thought `|` is functionally equivalent to `or`

Reply:
then it's a bit wired, and I need to take back what I said before 😅

Reply:
What matters here is how pandas.Series' `|` operator work.

Reply:
but the point is `|` is native python thing, not pandas thing.

In pandas, you could do `s1.eq(s2)` to compare, in the case @ueshin showed above, this is then
```python
>>>s1 = pd.Series([True, True, np.nan])
>>>s2 = pd.Series([True, np.nan, True])
>>>s1.eq(s2)
0     True
1    False
2    False
dtype: bool
```

Reply:
maybe worth reporting an issue in pandas 😄

Reply:
@ueshin @charlesdong1991 

https://github.com/pandas-dev/pandas/blob/master/pandas/tests/series/test_operators.py#L447-L452

```python
# True | np.nan => True
exp_or1 = pd.Series([True, True, True, False], index=list("ABCD"), name="x")
tm.assert_series_equal(s1 | s2, exp_or1)
# np.nan | True => np.nan, filled with False
exp_or = pd.Series([True, True, False, False], index=list("ABCD"), name="x")
tm.assert_series_equal(s2 | s1, exp_or)
```

Reply:
ahh, then it behaves this way on purpose in pandas. thanks for looking at source code to clear things up @harupy

Reply:
Let's discuss `&` and `|` and address if needed in a separate issue/PR.

Reply:
I think `&` and `|` should be addressed in a separate PR.

Reply:
Seems like pandas tries to follow the boolean operations rules in Python (`or` and `and`) - https://docs.python.org/3/library/stdtypes.html#boolean-operations-and-or-not

```
# True | np.nan => True
# np.nan | True => np.nan
```

Then, ^ could make sense.

However, I am still kind of lost about "np.nan, filled with False" because `bool(np.nan)` is coerced to `True`, e.g.)

```python
>>> pd.Series([float("nan")]).astype("bool")
0    True
dtype: bool
```